src/plugins/crypto/gpg.c

```diff
@@ -748,7 +748,8 @@ int ELEKTRA_PLUGIN_FUNCTION (gpgCall) (KeySet * conf, Key * errorKey, Key * msgK
 
 	case 1:
 		// bad signature
-		ELEKTRA_SET_ERROR (ELEKTRA_ERROR_CRYPTO_GPG, errorKey, "GPG reported a bad signature");
+		// TODO: Correct??
+		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "GPG reported a bad signature. errno: %s", strerror (errno));
```

Comment:
I think "Bad Signature" would mean here that maybe something is signed with a wrong key? I think semantic validation would fit best here as the syntactic signature will be correct I assume.

Reply:
Yes, but it could also be that someone modifies the config file and because of this the signature does not match anymore. Or someone modified the signature. Or maybe simply a bit flipped because of the hardware. In any case: yes, validation makes also sense to me.

Reply:
I see, so it could be both depending on the concrete errno I guess. I will still use semantic validation though for the most likely event.

Reply:
> I see, so it could be both depending on the concrete errno I guess.

Is there an errno set. If there is more information (which makes differences in the categorization), you should use this information.

Usually, however, the categories should be broad enough so that it does not matter.

> I will still use semantic validation though for the most likely event.

I am afraid that "semantic validation" is your catch all. It gets more and more vague what "semantic validation" is for you.

Reply:
>Is there an errno set. If there is more information (which makes differences in the categorization), you should use this information.
>Usually, however, the categories should be broad enough so that it does not matter.

I agree with you that it should be broad enough, but in this case where one error can be put into two sections depending on the errno it is not feasible unless you put it in a very broad category which then loses the benefit of categorization...

> I am afraid that "semantic validation" is your catch all. It gets more and more vague what "semantic validation" is for you.

That is funny that you think that `semantic validation` is my catch-all category. Actually it is the first or second time I used it where I am not 100% sure.

Reply:
Nonetheless, I do not assume that a hardware bit flip will be the standard case but a tampered signature or one that does not match anymore because of expiration. In the latter two cases semantic validation is definitely correct as syntactically everything is fine. Also the reaction of a semantic validation here is correct: retry but with a correct semantic value (use correct signing key, get a non-expired key and sign again)

Reply:
> That is funny

Then we do not need to worry :smile: 

> In the latter two cases semantic validation is definitely correct as syntactically everything is fine.

Ok, perfect then. Franz also agrees that semantic validation makes sense in this case (he wrote a paper about that topic). So you can also mention this in your thesis (maybe as example of what semantic validation is: signatures is definitely an interesting case).

We also spoke about your enums. So if you have an enum "red", "blue", and "green" then "violet" would be a semantic error but "redd" would be a syntactic error. This again is not possible to distinguish, so also here semantic error makes sense.